### DISCUSSION TITLE
Text on a lateral surface of a cylinder

### DISCUSSION THREAD

**Discussion by Paul8043** [gumyr/build123d#282] [Q&A] [answered]
Would like to extrude some text on a non-planar plane. Text should go the lateral surface of a cylinder. How to do?
```
m = Part()+Cylinder(10,20)
f = m.faces().sort_by(Axis.Z)
bot = f[0]
lat = f[1]              # lateral surface
top = f[2]
txt = Text("TEST",10)   # should be extruded on face "lat"
```

**Comment by gumyr** [comment]
I had to fix up some broken code so refresh your build123d install, but here is how one does this:
```py
from build123d import *
from ocp_vscode import *

r = 10
m = Cylinder(r, 20)
f = m.faces().sort_by(Axis.Z)
bot = f[0]
lat = f[1]  # lateral surface
top = f[2]
txt = Text("TEST", 10)  # should be extruded on face "lat"
path = Edge.make_circle(r)
txt_on_surface = lat.project_faces(faces=txt, path=path, start=0)
three_d_text = [face.thicken(1) for face in txt_on_surface.faces()]
show(lat, three_d_text, alphas=[0.2, 1])
```
![image](https://github.com/gumyr/build123d/assets/85638518/99213f8f-83d7-4987-a61f-c98bb1bcbb64)

To place text (or any faces) on a non-planar surface one needs to define the `path` that the text will follow which in this case is just a circle. The `project_faces` method uses  the shape (as self), the faces, the path, and a relative start point on the path (from 0.0 to 1.0) to place the planar text onto a plane tangent to the Shape at a point on the path, project it onto the Shape and return a list of projected Faces.  Given that the projected Faces are non-planar and wrapped around the cylinder I've used `thicken` here to make them into Solids not extrude which only works in a single direction.

**Reply by Paul8043** [reply, marked as answer]
Works perfectly, exactly what I needed. Thanks a lot. 

I am deeply impressed what Build123d can do. The potential is enormous!
